designs/2024-hooks-for-test-cases/README.md

```diff
@@ -0,0 +1,311 @@
+    - `setup` is for **specific** actions applicable to **particular** test.
+- Limited Use Case:
+  - The necessity for the `setup` might be limited to specific rules and plugins. The broader ESLint user
+    community may not frequently encounter these challenges, and it may seem unjustified to them.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+Adding new optional property to the test case is backwards compatible.
+When the property is not set, `RuleTester` acts the same for existing users — no disruption expected.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+My current workaround is assigning a custom `it` function to the `RuleTester` that looks up for a mock by test name:
+
+```javascript
+const readerMock = jest.fn();
+jest.mock('fs', () => ({
+  readFileSync: readerMock
+}));
+
+const mockedEnvironments = {
+  "test case one": { dependencies: { "some-package": "^1.0.0" } },
+  "test case two": { devDependencies: { "another-package": "^2.0.0" } }
+}
+
+RuleTester.it = (name, ...rest) => {
+  readerMock.mockReturnValue(
+    JSON.stringify( mockedEnvironments[name] )
+  );
+  it(name, ...rest);
+};
+
+new RuleTester().run("my-custom-rule", myCustomRule, {
+  valid: [
+    {
+      name: "test case one",
+      code: '/* valid code example */',
+      options: [/* some options */]
+    }
+  ],
+  invalid: [
+    {
+      code: '/* invalid code example */',
+      errors: [{ messageId: "someErrorId" }],
+      options: [/* some options */]
+    }
+  ]
+});
+```
+
+However, I'm not satisfied with this approach due to a number of disadvantages:
+
+- The `name` property is required to be set on each test case so that `it` could distinguish between them;
+- Also, that `name` must be unique for proper lookup;
+- And most importantly: mocked environment is detached from the test cases definitions:
+  - They are having only that weak reference via the `name`;
+  - It's harder to read, to maintain and to make some adjustments;
+  - The mocked environment is, in essence, one of the arguments to the rule being tested,
+    therefore it does make sense for them to be near `options` and `code`.
+
+## Open Questions
+
+<!--
+    This section is optional, but is suggested for a first draft.
+
+    What parts of this proposal are you unclear about? What do you
+    need to know before you can finalize this RFC?
+
+    List the questions that you'd like reviewers to focus on. When
+    you've received the answers and updated the design to reflect them, 
+    you can remove this section.
+-->
+
+- Should there also be the `teardown` property?
```

Comment:
I would say a way to teardown a test is warranted, as you don't to pollute the other tests (be they from ESLint or not).

Alternatively, the `setup` function could return an optional function to do the teardown. This could be useful if you need to reference something created in the ` setup` function?

Reply:
`setup` optionally returning a cleanup function, I like it.
Very similar to `React.useEffect()`.

Reply:
I think we do need a `teardown` equivalent, and I'd say, maybe we'd even want to require it to make sure other tests aren't polluted?

I'd much prefer a `teardown` method than returning a function from `setup`, as I think the latter is more easily forgotten.

Reply:
I also think we do need to support specifying a teardown method (which would be named `after()` since we've agreed on `before`/`after` naming in https://github.com/eslint/rfcs/pull/122#discussion_r1721948622 discussion).

And I think we should make sure this method is called even if the test fails.